src/lambda/wizards/samDeployWizard.ts

```diff
@@ -357,23 +383,84 @@ export class DefaultSamDeployWizardContext implements SamDeployWizardContext {
                     'AWS.samcli.deploy.s3Bucket.prompt',
                     'Enter the AWS S3 bucket to which your code should be deployed'
                 ),
+                value: initialValue || '',
+                matchOnDetail: true,
                 ignoreFocusOut: true,
-                prompt: localize(
-                    'AWS.samcli.deploy.s3Bucket.region',
-                    'S3 bucket must be in selected region: {0}',
-                    selectedRegion
-                )
-            }
+            },
+            // preloaded "busy screen" message
+            // using a const in case we want to make this clickable as a way to abort the lookup.
+            items: [
+                {
+                    label: loadingBuckets
+                }
+            ],
         })
 
-        // Pre-populate the value if it was already set
-        if (initialValue) {
-            inputBox.value = initialValue
+        // throw up a busy screen while the S3 buckets are loading
+        quickPick.enabled = false
+        quickPick.busy = true
+        quickPick.show()
+
+        // sort through buckets and add to cache if you had to run a listBuckets
+        if (listBucketsResponse && listBucketsResponse.Buckets) {
+            // listBuckets lists all buckets tied to an account regardless of region
+            // we need to filter the list ourselves since we do care about the region
+            const promises = listBucketsResponse.Buckets
+                .filter(bucket => bucket.Name)
+                .map(async bucket => {
+                    const bucketLocation = await s3Client.getBucketLocation(bucket.Name!)
+                    // intermediate var used to correct blank and 'EU' locations:
+                    // https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
+                    let transformedLocation = bucketLocation.LocationConstraint
+                    if (!transformedLocation || transformedLocation === '') {
+                        transformedLocation = 'us-east-1'
+                    } else if (transformedLocation === 'EU') {
+                        transformedLocation = 'eu-west-1'
+                    }
+                    this.addBucketToRegionMap(bucket.Name!, transformedLocation)
+                })
+
+            await Promise.all(promises)
         }
 
-        return await input.promptUser({
-            inputBox: inputBox,
-            onValidateInput: validateS3Bucket,
+        // does the selected region have buckets? If so, add to the QP
+        if (this.regionBucketMap.size > 0) {
```

Comment:
This can be made into a function too.
`const bucketItems = getBucketItemsForRegion(...)`

Reply:
The function could likely handle the logic for returning a stub item if the list is empty.